Add zffs

findlayfeng opened the pull request:
ZFFS is supposed to be used instead of NFFS.

Reason for creatin another FS is that there are so many serious bugs in NFFS.
While trying to fix them, I found that I was almost going to refactored it.
Why not rewrite one?

It has been roughly completed.

This PR contain the [flash_map external extension PR](https://github.com/zephyrproject-rtos/zephyr/pull/8837)
Since I used the modified flash_map api in #8837, I merged #8837 into it.

Unfinished work.

- [ ] Garbage collection
- [ ] Document
- [ ] Tests

I think I need some help.

Commit:
dts: flash_map: Make dts automatically generate a list of flash devices

Add zephyr,flash_map to choose, accept values or lists,
default is zephyr,flash
zephyr,flash_map is used to generate flash_drivers_map

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
dts: flash_map: Make dts automatically generate a list of flash area

generate default_flash_map

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
subsys: mcuboot: Introduces flash_map subsystem

It introduces flash_map subsystem to operate on flash
footprint instead of flash_driver.
Use flash_area ID index image, not offset.
Select FLASH_MAP.
Update tests.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
subsys: flash_img: Involvew flash map

Now flash img uses the flash area id to determine which flash partition
is being operated.
Update tests.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
ext: img_mgmt: update api about dfu

update flash img api
update mcuboot api

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
subsys: usb_dfu: update api about dfu

update flash img api
update mcuboot api

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
flash_map: Add some functions to get the sector information

flash_area_get_sector_info_by_offs(),
flash_area_get_sector_info_by_idx() Get sector information
flash_area_get_sector_count() Get sector information

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
subsys: fs: add zffs

todo

Signed-off-by: Findlay Feng <i@fengch.me>

hongshui3000 commented:
is it appropriate to put it in the ext folder?

hongshui3000 commented on subsys/fs/Kconfig:
```diff
@@ -44,6 +44,16 @@ config FILE_SYSTEM_NFFS
+	select FS_FLASH_STORAGE_PARTITION
+	select FLASH_MAP
+	help
+	  Enables NFFS file system support.
```
NFFS ->ZNFFS ?

findlayfeng commented:
@hongshui3000 lib folder ?

hongshui3000 commented:
There are many bugs in nffs.     Compared with nffs, what are the improvements of znffs? @findlayfeng

nvlsianpu commented:
> There are many bugs in nffs.

That is know - I have in my backlog to fix them all.

> Trying to fix it, I found that I almost refactored it.

I agree that there is a lot of effort required to fix the bugs, but not really refactoring 100% refactoring. Rather something like 20%. Two solution proposals were discuses here: https://github.com/apache/mynewt-nffs/issues/10

> Why not rewrite one?

We know that NFFS have two serious issue. many other bugs were squashed so far. Another FS will have similar childhood diseases. I prefer to fix NFFS than introduce another FS - even if this mean to refactor a lot (conclusion from discussion mentioned above was that we expected to refactor a lot).
I expected a lot of bug squashing for znffs as well - this is just by my experience. I saw similar situation many times - the last was NVS. Even tough someone  is a excellent developer - nobody is perfect.

But - adding a FS might have sense...
I can see sense in introduction of another FS if it give us serious capabilities which are not provided by any other FS we have. Adding a new module mean to maintain it as well. (This was the case when we accepted NVS introduction a few months ago)

I'm not against this path - but I need serious reasons, even for reviewing this.

@carlescufi ^^

findlayfeng commented:
嗯，采用nffs的一个理由是较少的ram占用。这个在小容量设备非常致命。
我是用的是 nrf51822_qfac(32kb ram)。
我仔细的阅读过nffs的代码。
对于内存的使用没有很好的控制，或者说不能以牺牲速度的前提下，节约内存的使用。
另外遇到的问题还有，以回收节点为目标的垃圾回收过程不是有效的，会造成一直不间断的copy
当几乎满节点的情况下，重建过程中遇到删除的节点不能正确处理，因为会超出系统的内存最大值。
节点数太容易达到上限（几乎所有操作都会增加节点，意味着更多的内存占用，为了合并更多的节点，垃圾回收需要更多的内存，非常的矛盾）
还有一个不够严谨的部分，crc，我认为存储结构设计者并不是非常的了解这种算法的特性
From Google Translate
Well, one reason to adopt nffs is that less ram is used. This is very deadly in small capacity devices.
I am using nrf51822_qfac (32kb ram).
I have read the code of nffs carefully.
There is no good control over the use of memory, or the use of memory can be saved without sacrificing speed.
Another problem encountered is that the garbage collection process targeting the recycling node is not effective and will result in uninterrupted copying.
In the case of almost full nodes, nodes that encounter deletion during the rebuild process cannot be processed correctly because the system's memory maximum is exceeded.
The number of nodes is too easy to reach the upper limit (almost all operations will increase the number of nodes, meaning more memory usage, in order to merge more nodes, garbage collection requires more memory, very contradictory)
There is also a less rigorous part, crc, I think storage architecture designers are not very aware of the characteristics of this algorithm.

findlayfeng commented:
@nvlsianpu nffs在内存中维护一个哈希表，所有的节点都会被添加到哈希表，这是一个在小内存的情况下限制nffs性能的因素。
我使用一个b-tree(增量式写入硬盘+缓存方式)的实现代替了它。
最后导致几乎需要重构，干脆完全的重构了。
From google translate
Nffs maintains a hash table in memory, and all nodes are added to the hash table, which is a factor limiting the performance of nffs in the case of small memory.
I replaced it with an implementation of a b-tree (incremental write to hard disk + cache).
In the end, it almost requires refactoring, and it is completely refactored.

findlayfeng commented:
@nvlsianpu I started to pay attention to the problem of nffs. The reason is what you discussed.
At first I thought it was wrong to call the delete, but also tried to patch, after a few tests and found that this is not correct, began to read the nffs code. More and more problems are starting to appear, and there are many fixes for problems in my local repository.
So you saw the prototype of znffs.
Zephyr api is used, so ‘z’ is added before the name

carlescufi commented:
@findlayfeng thanks for this submission.
The first thing to do in this PR is that if you are going to replace NFFS with a fork (znffs), then you should do the following things in this PR:

* Rename `znffs` to `zffs`
* Move the code from `ext/fs/znffs` to `subsys/fs/zffs`
* Make both filesystem available for a while, leaving `ext/fs/nffs` and `subsys/fs/nffs_fs.c` untouched

That way we will be able to compare and have both side-by-side until we decide to take one direction.

carlescufi commented:
@findlayfeng also please address the outstanding issues in #8837 so we can merge it, then this PR will not have any dependencies.

nvlsianpu commented:
I think what @carlescufi proposed (have nffs and znffs side by side as transient steep) have sense (discussed with him off- channel).
@findlayfeng Your proposition of implements algorithms differently also make sense (thanks for elaboration). Looks like making zffs will be better idea than fixing nffs (we saw that it was had to have support for it very recently)

I look roughly trough the code. At the first sight I saw that you should use zephyr primitives instead of own implementations - `misc/dlist.h`, `misc/slist.h`.

findlayfeng commented:
> I think what @carlescufi proposed (have nffs and znffs side by side as transient steep) have sense (discussed with him off- channel).

Thanks for your proposed, I will adjust the name

> @findlayfeng Your proposition of implements algorithms differently also make sense (thanks for elaboration). Looks like making zffs will be better idea than fixing nffs (we saw that it was had to have support for it very recently)
> 
> I look roughly trough the code. At the first sight I saw that you should use zephyr primitives instead of own implementations - `misc/dlist.h`, `misc/slist.h`.

Thanks for your advice.
The slist and dlist here are implemented on disk and not memory.
I think `misc/list_gen.h` is probably used, unlike the peek function of `misc/dlist.h`, `misc/slist.h`.

Commit:
dts: mtd: Refine flash classification

Different flash devices will rely on different buses,
so use different .yaml files.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
drivers: flash: add Nordic qspi flash controller.

Qspi with nrf52840 enabled and Added driver code(block only).

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
samples: drivers: add qspi_flash_nrf.

Updated the dts configuration of pca10056.
Added an example of qspi flash controller.

Signed-off-by: Findlay Feng <i@fengch.me>

Commit:
samples: flash_map: add sample for nrf52840_pca10056

Added flash_map example

Signed-off-by: Findlay Feng <i@fengch.me>

hongshui3000 commented on subsys/settings/Kconfig:
```diff
@@ -6,9 +6,9 @@
 menuconfig SETTINGS
 	bool "Enable settings subsystem with non-volatile storage"
-	# Only NFFS is currently supported as FS.
+	# NFFS and ZNFFS is currently supported as FS.
```
ZNFFS -->ZFFS ?

carlescufi replied:
Replace with `ZFFS`

hongshui3000 commented on subsys/settings/Kconfig:
```diff
@@ -6,9 +6,9 @@
+	# NFFS and ZNFFS is currently supported as FS.
 	# The reason in that FatFs doesn't implement the fs_rename() API
-	depends on (FILE_SYSTEM && FILE_SYSTEM_NFFS) || (FCB && FLASH_PAGE_LAYOUT)
+	depends on (FILE_SYSTEM && (FILE_SYSTEM_NFFS || FILE_SYSTEM_ZNFFS)) || (FCB && FLASH_PAGE_LAYOUT)
```
FILE_SYSTEM_ZNFFS --> FILE_SYSTEM_ZFFS  ?

carlescufi replied:
Replace with `ZFFS`

carlescufi commented on subsys/fs/CMakeLists.txt:
```diff
@@ -2,18 +2,22 @@ if(CONFIG_FILE_SYSTEM)
   zephyr_interface_library_named(FS)
   zephyr_link_interface_ifdef(CONFIG_FAT_FILESYSTEM_ELM ELMFAT)
   zephyr_link_interface_ifdef(CONFIG_FILE_SYSTEM_NFFS   NFFS)
+  zephyr_link_interface_ifdef(CONFIG_FILE_SYSTEM_ZNFFS  ZNFFS)
```
Replace with `ZFFS`

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+			}
+			key_count++;
+		}
+	} while (rc != -ENOENT);
```
what is rc == -EIO? Aren't you going to loop forever here?

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+	do {
+		addr = zffs_area_pointer_to_addr(zffs, &pointer);
+		rc = zffs_object_check(zffs, &pointer, &id);
+		if (rc == 0) {
```
shouldn't you bail out here as soon as rc != 0? (see my comment below)

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+		switch (data->type) {
+		case ZFFS_TYPE_DIR:
+			SYS_SLIST_FOR_EACH_NODE(&zffs->opened, sn)
+			{
```
SYS_SLIST_FOR_EACH_NODE() is actually creating a for loop, so you need to put the '{' on the same line 
(that rule stands for if/for/while/switch). Apply that everywhere where relevant

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+			return rc;
+		}
+	}
+	else {
```
else should be on same line as previous }, apply that everywhere where relevant.

findlayfeng replied:
yep

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+					   &pointer);
+		rc = zffs_object_check(zffs, &pointer, &id);
+		if (rc) {
+			// 删除这个键 地址块被遗弃
```
Commenting in english (and with /**/, not //) would be better for all the contributors ;)

findlayfeng replied:
This is a todo reserved for myself.
It will disappear when it is finished

tbursztyka commented on subsys/fs/zffs/src/zffs_misc.c:
```diff
@@ -0,0 +1,193 @@
+	int lock = k_mutex_lock(&zffs->lock, K_NO_WAIT);
+	int id;
+
+	if (zffs->next_id == ZFFS_ROOT_ID) {
```
So getting the lock here isn't a condition to actually temper zffs content? Worth verifying this.

findlayfeng replied:
Now I think this lock can be deleted.
When adding this function, it was considered that it might be called outside the zffs environment.
For now, this is unlikely.

tbursztyka commented on subsys/fs/zffs/src/zffs_dlist.c:
```diff
@@ -0,0 +1,252 @@
+	if (rc) {
+		return rc;
+	}
+	if (head.object_type != ZFFS_OBJECT_TYPE_DLIST_NODE) {
```
have an empty line after a } (unless it's an else or another }) Apply that everywhere where relevant

findlayfeng replied:
yep

tbursztyka commented on subsys/fs/zffs/src/zffs_dir.c:
```diff
@@ -0,0 +1,699 @@
+
+		return 0;
+	}
+	else {
```
the else {} here is unnecessary, it's obvious that if the condition above fails it will end up here (since you always return from the if () {} also)

findlayfeng replied:
yep

tbursztyka commented on subsys/fs/zffs/src/zffs_dir.c:
```diff
@@ -0,0 +1,699 @@
+	return 0;
+}
+
+// int zffs_dir_append(struct zffs_data *zffs,
```
use /**/, not //

findlayfeng replied:
Already not needed, forget to clean up when refactoring

tbursztyka commented on subsys/fs/zffs/src/zffs_file.c:
```diff
@@ -0,0 +1,456 @@
+{
+	int rc;
+
+	if (file_is_write(file)) {
```
change this to:
if  (!file_is_write(file)) { 
return 0;
}

Code below won't be that much indented then. easier to read

findlayfeng replied:
yep

tbursztyka commented on subsys/fs/zffs/src/zffs_file.c:
```diff
@@ -0,0 +1,456 @@
+					       sizeof(disk));
+		} else {
+			rc = zffs_dlist_updata_ex(
+				zffs, pointer, file->node.id, &disk, sizeof(disk));
```
you could have put zffs, pointer in the beginning of parameter list, etc...

findlayfeng replied:
yep

tbursztyka commented on subsys/fs/zffs/src/zffs_file.c:
```diff
@@ -0,0 +1,456 @@
+			}
+
+			rc =
+				zffs_object_delete(zffs, pointer, file->block.id);
```
? 
coalesce with line 159

tbursztyka reviewed (changes requested):
Verify the code style.

Also, in some places there are so many if/for/do/while/switch imbricated (4 or more) that it makes the code hard to read. Usually this can be improved a lot by reversing and exiting from some test, or moving parts of code in dedicated functions, or even using carefully goto.

findlayfeng commented on subsys/fs/zffs/src/zffs_object.c:
```diff
@@ -0,0 +1,177 @@
+			     struct object_foreach_data *data)
+{
+	int rc = object_open(zffs, data->pointer, key, value);
+	if (rc < 0) {
```
Regarding the code style, when writing quickly, I don’t pay attention to it many times, all the style problems will be adjusted.

hongshui3000 commented:
What is the status of this pull request?

nvlsianpu commented:
looks like orphaned.

nvlsianpu commented:
I'm  closing this PR as:
- PR is orphaned by the contributor, no others interested in finishing it
- litlefs was introduced as another substitute to NFFS. It is much better to use it.
- the code is far from being ready to merge (even without conflicts)
